### ISSUE TITLE
[Win32] Korean input behaviour is different than native OS

### ISSUE THREAD

**Issue by yunu7067** [flutter/flutter#88645]
There is no problem in English. However, there are some problems related to 'space' key in Korean. If I press space only once, a key such as Ctrl + A or Backspace will not work. In that state, if I press the space one more time, the space will be entered and the keys mentioned above will work. Korean alphabet(Hangul) are entered normally with only one press on the space. 

![test](https://user-images.githubusercontent.com/57068092/130330036-97888bc7-4a8a-4e03-8274-555352c2ffaf.gif)

This is a picture of the error captured. The key I pressed is displayed in teh lower left corner. Space, Backspace, and Delete key order.

My env
* CPU : AMD Ryzen 7 1700X
* VGA : AMD Radeon RX 580X
* OS : Windows 10 Home 64bit (build 19043)
* IDE : Jetbrains IntelliJ IDEA 2021.2 Ultimate Editoin

<details>
  <summary>main.dart</summary>

```
import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: Scaffold(
        body: Container(
          child: TextField(),
        )
      )
    );
  }
}

```

</details>

<details>
  <summary>Logs</summary>

```
[  +87 ms] executing: [C:\Users\robin\Documents\flutter/] git -c log.showSignature=false log -n 1
--pretty=format:%H
[ +149 ms] Exit code 0 from: git -c log.showSignature=false log -n 1 --pretty=format:%H
[   +1 ms] f4abaa0735eba4dfd8f33f73363911d63931fe03
[        ] executing: [C:\Users\robin\Documents\flutter/] git tag --points-at
f4abaa0735eba4dfd8f33f73363911d63931fe03
[  +47 ms] Exit code 0 from: git tag --points-at f4abaa0735eba4dfd8f33f73363911d63931fe03
[        ] 2.2.3
[   +6 ms] executing: [C:\Users\robin\Documents\flutter/] git rev-parse --abbrev-ref --symbolic @{u}
[  +36 ms] Exit code 0 from: git rev-parse --abbrev-ref --symbolic @{u}
[        ] origin/stable
[        ] executing: [C:\Users\robin\Documents\flutter/] git ls-remote --get-url origin
[  +39 ms] Exit code 0 from: git ls-remote --get-url origin
[        ] https://github.com/flutter/flutter.git
[ +298 ms] Unable to locate an Android SDK.
[   +7 ms] executing: [C:\Users\robin\Documents\flutter/] git rev-parse --abbrev-ref HEAD
[  +33 ms] Exit code 0 from: git rev-parse --abbrev-ref HEAD
[   +1 ms] stable
[  +73 ms] Artifact Instance of 'AndroidGenSnapshotArtifacts' is not required, skipping update.
[   +1 ms] Artifact Instance of 'AndroidInternalBuildArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'IOSEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'FlutterWebSdk' is not required, skipping update.
[   +3 ms] Artifact Instance of 'WindowsEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'MacOSEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'LinuxEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'LinuxFuchsiaSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'MacOSFuchsiaSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'FlutterRunnerSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'FlutterRunnerDebugSymbols' is not required, skipping update.
[ +110 ms] Artifact Instance of 'AndroidGenSnapshotArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'AndroidInternalBuildArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'IOSEngineArtifacts' is not required, skipping update.
[   +5 ms] Artifact Instance of 'MacOSEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'LinuxEngineArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'LinuxFuchsiaSDKArtifacts' is not required, skipping update.
[        ] Artifact Instance of 'MacOSFuchsiaSDKArtifacts' is not required, skipping update.
[   +1 ms] Artifact Instance of 'FlutterRunnerSDKArtifacts' is not required, skipping update.
[   +1 ms] Artifact Instance of 'FlutterRunnerDebugSymbols' is not required, skipping update.
[  +94 ms] Skipping pub get: version match.
[ +133 ms] Generating
C:\Users\robin\Documents\workspace\twoeat_flutter_demo\android\app\src\main\java\io\flutter\plugins\GeneratedP
luginRegistrant.java
[ +112 ms] Initializing file store
[  +11 ms] Skipping target: gen_localizations
[   +8 ms] complete
[   +5 ms] Launching lib\main.dart on Windows in debug mode...
[   +6 ms] C:\Users\robin\Documents\flutter\bin\cache\dart-sdk\bin\dart.exe --disable-dart-dev
C:\Users\robin\Documents\flutter\bin\cache\artifacts\engine\windows-x64\frontend_server.dart.snapshot
--sdk-root C:\Users\robin\Documents\flutter\bin\cache\artifacts\engine\common\flutter_patched_sdk/
--incremental --target=flutter --debugger-module-names --experimental-emit-debug-metadata
-DFLUTTER_WEB_AUTO_DETECT=true --output-dill
C:\Users\robin\AppData\Local\Temp\flutter_tools.c0f86ca2\flutter_tool.cede4ffd\app.dill --packages
C:\Users\robin\Documents\workspace\twoeat_flutter_demo\.dart_tool\package_config.json -Ddart.vm.profile=false
-Ddart.vm.product=false --enable-asserts --track-widget-creation --filesystem-scheme org-dartlang-root
--initialize-from-dill build\3c113a45063dc6628e68a4111abcacad.cache.dill.track.dill
--enable-experiment=alternative-invalidation-strategy
[  +27 ms] executing: C:\Program Files (x86)\Microsoft Visual Studio\Installer\vswhere.exe -format json
-products * -utf8 -latest -version 16 -requires Microsoft.VisualStudio.Workload.NativeDesktop
Microsoft.VisualStudio.Component.VC.Tools.x86.x64 Microsoft.VisualStudio.Component.VC.CMake.Project
[  +38 ms] Exit code 0 from: C:\Program Files (x86)\Microsoft Visual Studio\Installer\vswhere.exe -format json
-products * -utf8 -latest -version 16 -requires Microsoft.VisualStudio.Workload.NativeDesktop
Microsoft.VisualStudio.Component.VC.Tools.x86.x64 Microsoft.VisualStudio.Component.VC.CMake.Project
[   +2 ms] [
             {
               "instanceId": "f6a5429f",
               "installDate": "2021-08-20T10:54:09Z",
               "installationName": "VisualStudio/16.11.1+31613.86",
               "installationPath": "C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\Community",
               "installationVersion": "16.11.31613.86",
               "productId": "Microsoft.VisualStudio.Product.Community",
               "productPath": "C:\\Program Files (x86)\\Microsoft Visual
               Studio\\2019\\Community\\Common7\\IDE\\devenv.exe",
               "state": 4294967295,
               "isComplete": true,
               "isLaunchable": true,
               "isPrerelease": false,
               "isRebootRequired": false,
               "displayName": "Visual Studio Community 2019",
               "description": "강력한 IDE, 학생, 오픈 소스 제공자 및 개인을 위해 무료로 제공",
               "channelId": "VisualStudio.16.Release",
               "channelUri": "https://aka.ms/vs/16/release/channel",
               "enginePath": "C:\\Program Files (x86)\\Microsoft Visual
               Studio\\Installer\\resources\\app\\ServiceHub\\Services\\Microsoft.VisualStudio.Setup.Service",
               "releaseNotes":
               "https://docs.microsoft.com/en-us/visualstudio/releases/2019/release-notes-v16.11#16.11.1",
               "thirdPartyNotices": "https://go.microsoft.com/fwlink/?LinkId=660909",
               "updateDate": "2021-08-20T10:54:09.8669086Z",
               "catalog": {
                 "buildBranch": "d16.11",
                 "buildVersion": "16.11.31613.86",
                 "id": "VisualStudio/16.11.1+31613.86",
                 "localBuild": "build-lab",
                 "manifestName": "VisualStudio",
                 "manifestType": "installer",
                 "productDisplayVersion": "16.11.1",
                 "productLine": "Dev16",
                 "productLineVersion": "2019",
                 "productMilestone": "RTW",
                 "productMilestoneIsPreRelease": "False",
                 "productName": "Visual Studio",
                 "productPatchVersion": "1",
                 "productPreReleaseMilestoneSuffix": "1.0",
                 "productSemanticVersion": "16.11.1+31613.86",
                 "requiredEngineVersion": "2.11.35.61819"
               },
               "properties": {
                 "campaignId": "",
                 "channelManifestId": "VisualStudio.16.Release/16.11.1+31613.86",
                 "nickname": "",
                 "setupEngineFilePath": "C:\\Program Files (x86)\\Microsoft Visual
                 Studio\\Installer\\setup.exe"
               }
             }
           ]
[   +4 ms] Building Windows application...
[  +16 ms] <- compile package:twoeat_flutter_demo/main.dart
[   +3 ms] executing: C:\Program Files (x86)\Microsoft Visual
Studio\2019\Community\Common7\IDE\CommonExtensions\Microsoft\CMake\CMake\bin\cmake.exe -S
C:\Users\robin\Documents\workspace\twoeat_flutter_demo\windows -B build\windows -G Visual Studio 16 2019
[  +53 ms] -- Selecting Windows SDK version 10.0.19041.0 to target Windows 10.0.19043.
[  +41 ms] -- Configuring done
[ +394 ms] -- Generating done
[  +82 ms] -- Build files have been written to:
C:/Users/robin/Documents/workspace/twoeat_flutter_demo/build/windows
[   +8 ms] executing: C:\Program Files (x86)\Microsoft Visual
Studio\2019\Community\Common7\IDE\CommonExtensions\Microsoft\CMake\CMake\bin\cmake.exe --build build\windows
--config Debug --target INSTALL --verbose
[ +162 ms] .NET Framework용 Microsoft (R) Build Engine 버전 16.11.0+0538acc04
[   +4 ms] Copyright (C) Microsoft Corporation. All rights reserved.
[ +241 ms] 빌드 시작: 2021-08-22 오전 1:55:30
[ +970 ms] 1 노드의 "C:\Users\robin\Documents\workspace\twoeat_flutter_demo\build\windows\INSTALL.vcxproj"
프로젝트(기본 대상)입니다.
[   +5 ms] "C:\Users\robin\Documents\workspace\twoeat_flutter_demo\build\windows\INSTALL.vcxproj"(1) 프로젝트
가 1
노드에서 "C:\Users\robin\Documents\workspace\twoeat_flutter_demo\build\windows\ZERO_CHECK.vcxproj"(2)을(를) 빌
드하고
있습니다(기본 대상).
[   +3 ms] InitializeBuildStatus:
[   +1 ms]   "AlwaysCreate"이(가) 지정되었기 때문에 "x64\Debug\ZERO_CHECK\ZERO_CHECK.tlog\unsuccessfulbuild"을
(를) 만들고
있습니다.
[  +47 ms] CustomBuild:
[   +1 ms]   모든 출력이 최신 상태입니다.
[  +37 ms] FinalizeBuildStatus:
[   +1 ms]   "x64\Debug\ZERO_CHECK\ZERO_CHECK.tlog\unsuccessfulbuild" 파일을 삭제하고 있습니다.
[        ]   "x64\Debug\ZERO_CHECK\ZERO_CHECK.tlog\ZERO_CHECK.lastbuildstate"에 연결(touching)하고 있습니다.
[   +3 ms] "C:\Users\robin\Documents\workspace\twoeat_flutter_demo\build\windows\ZERO_CHECK.vcxproj" 프로젝트
를
빌드했습니다(기본 대상).
[ +128 ms] "C:\Users\robin\Documents\workspace\twoeat_flutter_demo\build\windows\INSTALL.vcxproj"(1) 프로젝트
가 1
노드에서 "C:\Users\robin\Documents\workspace\twoeat_flutter_demo\build\windows\ALL_BUILD.vcxproj"(3)을(를) 빌
드하고
있습니다(기본 대상).
[   +2 ms] "C:\Users\robin\Documents\workspace\twoeat_flutter_demo\build\windows\ALL_BUILD.vcxproj"(3) 프로젝
트가 1
노드에서
"C:\Users\robin\Documents\workspace\twoeat_flutter_demo\build\windows\flutter\flutter_wrapper_app.vcxproj"(4)
을
(를) 빌드하고 있습니다(기본 대상).
[   +1 ms]
"C:\Users\robin\Documents\workspace\twoeat_flutter_demo\build\windows\flutter\flutter_wrapper_app.vcxproj"(4)
프로젝트가 1 노드에서
"C:\Users\robin\Documents\workspace\twoeat_flutter_demo\build\windows\flutter\flutter_assemble.vcxproj"(5)을(
를)
빌드하고 있습니다(기본 대상).
[   +1 ms] InitializeBuildStatus:
[        ]   "AlwaysCreate"이(가) 지정되었기 때문에
"x64\Debug\flutter_assemble\flutter_assemble.tlog\unsuccessfulbuild"을(를) 만들고 있습니다.
[   +9 ms] CustomBuild:
[   +1 ms]
"'C:\Users\robin\Documents\workspace\twoeat_flutter_demo\build\windows\CMakeFiles\706a7485052514f04372213d9b21
e2da\flutter_assemble.rule'의 빌드는
'C:\Users\robin\Documents\workspace\twoeat_flutter_demo\build\windows\CMakeFiles\053cc5b821ad6a281fd5a0be3ed75
d61\flutter_windows.dll.rule'의 빌드로 생성되는
'C:\USERS\ROBIN\DOCUMENTS\WORKSPACE\TWOEAT_FLUTTER_DEMO\WINDOWS\FLUTTER\EPHEMERAL\FLUTTER_WINDOWS.DLL'에 따라
달라집니다. 항목은 동시에 빌드할 수 없습니다."
[  +99 ms]   Generating
C:/Users/robin/Documents/workspace/twoeat_flutter_demo/windows/flutter/ephemeral/flutter_windows.dll,
C:/Users/robin/Documents/workspace/twoeat_flutter_demo/windows/flutter/ephemeral/flutter_export.h,
C:/Users/robin/Documents/workspace/twoeat_flutter_demo/windows/flutter/ephemeral/flutter_windows.h,
C:/Users/robin/Documents/workspace/twoeat_flutter_demo/windows/flutter/ephemeral/flutter_messenger.h,
C:/Users/robin/Documents/workspace/twoeat_flutter_demo/windows/flutter/ephemeral/flutter_plugin_registrar.h,
C:/Users/robin/Documents/workspace/twoeat_flutter_demo/windows/flutter/ephemeral/flutter_texture_registrar.h,
C:/Users/robin/Documents/workspace/twoeat_flutter_demo/windows/flutter/ephemeral/cpp_client_wrapper/core_imple
mentations.cc,
C:/Users/robin/Documents/workspace/twoeat_flutter_demo/windows/flutter/ephemeral/cpp_client_wrapper/standard_c
odec.cc,
C:/Users/robin/Documents/workspace/twoeat_flutter_demo/windows/flutter/ephemeral/cpp_client_wrapper/plugin_reg
istrar.cc,
C:/Users/robin/Documents/workspace/twoeat_flutter_demo/windows/flutter/ephemeral/cpp_client_wrapper/flutter_en
gine.cc,
C:/Users/robin/Documents/workspace/twoeat_flutter_demo/windows/flutter/ephemeral/cpp_client_wrapper/flutter_vi
ew_controller.cc, _phony_
[+1263 ms]   [  +90 ms] executing: [C:\Users\robin\Documents\flutter/] git -c log.showSignature=false log -n 1
--pretty=format:%H
[ +116 ms]   [ +142 ms] Exit code 0 from: git -c log.showSignature=false log -n 1 --pretty=format:%H
[   +1 ms]   [        ] f4abaa0735eba4dfd8f33f73363911d63931fe03
[        ]   [        ] executing: [C:\Users\robin\Documents\flutter/] git tag --points-at
f4abaa0735eba4dfd8f33f73363911d63931fe03
[  +66 ms]   [  +67 ms] Exit code 0 from: git tag --points-at f4abaa0735eba4dfd8f33f73363911d63931fe03
[   +1 ms]   [        ] 2.2.3
[   +6 ms]   [   +7 ms] executing: [C:\Users\robin\Documents\flutter/] git rev-parse --abbrev-ref --symbolic
@{u}
[  +62 ms]   [  +62 ms] Exit code 0 from: git rev-parse --abbrev-ref --symbolic @{u}
[   +1 ms]   [        ] origin/stable
[        ]   [        ] executing: [C:\Users\robin\Documents\flutter/] git ls-remote --get-url origin
[  +48 ms]   [  +50 ms] Exit code 0 from: git ls-remote --get-url origin
[   +1 ms]   [        ] https://github.com/flutter/flutter.git
[ +403 ms]   [ +403 ms] Unable to locate an Android SDK.
[  +11 ms]   [  +11 ms] executing: [C:\Users\robin\Documents\flutter/] git rev-parse --abbrev-ref HEAD
[  +51 ms]   [  +51 ms] Exit code 0 from: git rev-parse --abbrev-ref HEAD
[   +1 ms]   [        ] stable
[  +76 ms]   [  +76 ms] Artifact Instance of 'AndroidGenSnapshotArtifacts' is not required, skipping update.
[   +1 ms]   [        ] Artifact Instance of 'AndroidInternalBuildArtifacts' is not required, skipping update.
[        ]   [        ] Artifact Instance of 'IOSEngineArtifacts' is not required, skipping update.
[        ]   [        ] Artifact Instance of 'FlutterWebSdk' is not required, skipping update.
[   +2 ms]   [   +4 ms] Artifact Instance of 'WindowsEngineArtifacts' is not required, skipping update.
[        ]   [        ] Artifact Instance of 'MacOSEngineArtifacts' is not required, skipping update.
[        ]   [        ] Artifact Instance of 'LinuxEngineArtifacts' is not required, skipping update.
[        ]   [        ] Artifact Instance of 'LinuxFuchsiaSDKArtifacts' is not required, skipping update.
[        ]   [        ] Artifact Instance of 'MacOSFuchsiaSDKArtifacts' is not required, skipping update.
[        ]   [        ] Artifact Instance of 'FlutterRunnerSDKArtifacts' is not required, skipping update.
[        ]   [        ] Artifact Instance of 'FlutterRunnerDebugSymbols' is not required, skipping update.
[ +139 ms]   [ +142 ms] Artifact Instance of 'MaterialFonts' is not required, skipping update.
[   +1 ms]   [        ] Artifact Instance of 'GradleWrapper' is not required, skipping update.
[        ]   [        ] Artifact Instance of 'AndroidGenSnapshotArtifacts' is not required, skipping update.
[        ]   [        ] Artifact Instance of 'AndroidInternalBuildArtifacts' is not required, skipping update.
[        ]   [        ] Artifact Instance of 'IOSEngineArtifacts' is not required, skipping update.
[        ]   [        ] Artifact Instance of 'FlutterWebSdk' is not required, skipping update.
[        ]   [        ] Artifact Instance of 'FlutterSdk' is not required, skipping update.
[        ]   [   +2 ms] Artifact Instance of 'MacOSEngineArtifacts' is not required, skipping update.
[        ]   [        ] Artifact Instance of 'LinuxEngineArtifacts' is not required, skipping update.
[        ]   [        ] Artifact Instance of 'LinuxFuchsiaSDKArtifacts' is not required, skipping update.
[        ]   [        ] Artifact Instance of 'MacOSFuchsiaSDKArtifacts' is not required, skipping update.
[        ]   [        ] Artifact Instance of 'FlutterRunnerSDKArtifacts' is not required, skipping update.
[        ]   [        ] Artifact Instance of 'FlutterRunnerDebugSymbols' is not required, skipping update.
[        ]   [        ] Artifact Instance of 'IosUsbArtifacts' is not required, skipping update.
[        ]   [        ] Artifact Instance of 'IosUsbArtifacts' is not required, skipping update.
[        ]   [        ] Artifact Instance of 'IosUsbArtifacts' is not required, skipping update.
[        ]   [        ] Artifact Instance of 'IosUsbArtifacts' is not required, skipping update.
[   +1 ms]   [        ] Artifact Instance of 'IosUsbArtifacts' is not required, skipping update.
[        ]   [        ] Artifact Instance of 'FontSubsetArtifacts' is not required, skipping update.
[        ]   [        ] Artifact Instance of 'PubDependencies' is not required, skipping update.
[  +32 ms]   [  +38 ms] Initializing file store
[  +19 ms]   [  +19 ms] Done initializing file store
[  +49 ms]   [  +49 ms] Skipping target: gen_localizations
[+2690 ms]   [+2690 ms] Skipping target: unpack_windows
[ +466 ms]   [ +465 ms] Skipping target: kernel_snapshot
[   +5 ms]   [   +5 ms] invalidated build due to missing files:
C:\Users\robin\Documents\workspace\twoeat_flutter_demo\DOES_NOT_EXIST_RERUN_FOR_WILDCARD300975220
[ +241 ms]   [ +241 ms] debug_bundle_windows_assets: Starting due to {InvalidatedReasonKind.inputMissing: The
following inputs were missing:
C:\Users\robin\Documents\workspace\twoeat_flutter_demo\DOES_NOT_EXIST_RERUN_FOR_WILDCARD300975220}
[ +137 ms]   [ +137 ms] Manifest contained wildcard assets. Inserting missing file into build graph to force
rerun. for more information see #56466.
[  +26 ms]   [  +26 ms] debug_bundle_windows_assets: Complete
[ +246 ms]   [ +246 ms] Persisting file store
[  +13 ms]   [  +12 ms] Done persisting file store
[   +8 ms]   [   +8 ms] build succeeded.
[  +10 ms]   [  +11 ms] "flutter assemble" took 4,128ms.
[ +257 ms]   [ +257 ms] ensureAnalyticsSent: 252ms
[   +2 ms]   [   +1 ms] Running shutdown hooks
[   +1 ms]   [        ] Shutdown hooks complete
[        ]   [        ] exiting with code 0
[ +131 ms] FinalizeBuildStatus:
[   +1 ms]   "x64\Debug\flutter_assemble\flutter_assemble.tlog\unsuccessfulbuild" 파일을 삭제하고 있습니다.
[        ]   "x64\Debug\flutter_assemble\flutter_assemble.tlog\flutter_assemble.lastbuildstate"에
연결(touching)하고 있습니다.
[   +1 ms]
"C:\Users\robin\Documents\workspace\twoeat_flutter_demo\build\windows\flutter\flutter_assemble.vcxproj" 프로젝
트를
빌드했습니다(기본 대상).
[   +6 ms] InitializeBuildStatus:
[        ]   "AlwaysCreate"이(가) 지정되었기 때문에
"flutter_wrapper_app.dir\Debug\flutter_.18FDA4FB.tlog\unsuccessfulbuild"을(를) 만들고 있습니다.
[   +2 ms] CustomBuild:
[        ]   모든 출력이 최신 상태입니다.
[  +64 ms] ClCompile:
[        ]   모든 출력이 최신 상태입니다.
[  +26 ms] Lib:
[        ]   모든 출력이 최신 상태입니다.
[        ]   flutter_wrapper_app.vcxproj ->
C:\Users\robin\Documents\workspace\twoeat_flutter_demo\build\windows\flutter\Debug\flutter_wrapper_app.lib
[  +16 ms] FinalizeBuildStatus:
[   +1 ms]   "flutter_wrapper_app.dir\Debug\flutter_.18FDA4FB.tlog\unsuccessfulbuild" 파일을 삭제하고 있습니다
.
[        ]   "flutter_wrapper_app.dir\Debug\flutter_.18FDA4FB.tlog\flutter_wrapper_app.lastbuildstate"에
연결(touching)하고 있습니다.
[        ]
"C:\Users\robin\Documents\workspace\twoeat_flutter_demo\build\windows\flutter\flutter_wrapper_app.vcxproj"
프로젝트를 빌드했습니다(기본 대상).
[  +63 ms] "C:\Users\robin\Documents\workspace\twoeat_flutter_demo\build\windows\ALL_BUILD.vcxproj"(3) 프로젝
트가 1
노드에서
"C:\Users\robin\Documents\workspace\twoeat_flutter_demo\build\windows\flutter\flutter_wrapper_plugin.vcxproj"(
6)을(를) 빌드하고 있습니다(기본 대상).
[   +1 ms] InitializeBuildStatus:
[        ]   "AlwaysCreate"이(가) 지정되었기 때문에
"flutter_wrapper_plugin.dir\Debug\flutter_.15075795.tlog\unsuccessfulbuild"을(를) 만들고 있습니다.
[   +1 ms] CustomBuild:
[        ]   모든 출력이 최신 상태입니다.
[  +10 ms] ClCompile:
[        ]   모든 출력이 최신 상태입니다.
[   +4 ms] Lib:
[        ]   모든 출력이 최신 상태입니다.
[        ]   flutter_wrapper_plugin.vcxproj ->
C:\Users\robin\Documents\workspace\twoeat_flutter_demo\build\windows\flutter\Debug\flutter_wrapper_plugin.lib
[   +7 ms] FinalizeBuildStatus:
[        ]   "flutter_wrapper_plugin.dir\Debug\flutter_.15075795.tlog\unsuccessfulbuild" 파일을 삭제하고 있습
니다.
[        ]   "flutter_wrapper_plugin.dir\Debug\flutter_.15075795.tlog\flutter_wrapper_plugin.lastbuildstate"에
연결(touching)하고 있습니다.
[        ]
"C:\Users\robin\Documents\workspace\twoeat_flutter_demo\build\windows\flutter\flutter_wrapper_plugin.vcxproj"
프로젝트를 빌드했습니다(기본 대상).
[  +56 ms] "C:\Users\robin\Documents\workspace\twoeat_flutter_demo\build\windows\ALL_BUILD.vcxproj"(3) 프로젝
트가 1
노드에서
"C:\Users\robin\Documents\workspace\twoeat_flutter_demo\build\windows\runner\twoeat_flutter_demo.vcxproj"(7)을
(
를) 빌드하고 있습니다(기본 대상).
[   +2 ms] InitializeBuildStatus:
[        ]   "twoeat_flutter_demo.dir\Debug\twoeat_f.FA8FDEFE.tlog\unsuccessfulbuild"에 연결(touching)하고 있
습니다.
[   +1 ms] CustomBuild:
[        ]   모든 출력이 최신 상태입니다.
[  +31 ms] ClCompile:
[   +1 ms]   C:\Program Files (x86)\Microsoft Visual
Studio\2019\Community\VC\Tools\MSVC\14.29.30133\bin\HostX64\x64\CL.exe /c
/IC:\Users\robin\Documents\workspace\twoeat_flutter_demo\windows
/IC:\Users\robin\Documents\workspace\twoeat_flutter_demo\windows\flutter\ephemeral
/IC:\Users\robin\Documents\workspace\twoeat_flutter_demo\windows\flutter\ephemeral\cpp_client_wrapper\include
/Zi /nologo /W4 /WX /diagnostics:column /Od /Ob0 /D WIN32 /D _WINDOWS /D _HAS_EXCEPTIONS=0 /D _DEBUG /D
NOMINMAX /D UNICODE /D _UNICODE /D "CMAKE_INTDIR=\"Debug\"" /D _UNICODE /D UNICODE /Gm- /EHsc /RTC1 /MDd /GS
/fp:precise /Zc:wchar_t /Zc:forScope /Zc:inline /GR /std:c++17 /Fo"twoeat_flutter_demo.dir\Debug\\"
/Fd"twoeat_flutter_demo.dir\Debug\vc142.pdb" /external:W4 /Gd /TP /wd4100 /errorReport:queue
C:\Users\robin\Documents\workspace\twoeat_flutter_demo\windows\flutter\generated_plugin_registrant.cc
C:\Users\robin\Documents\workspace\twoeat_flutter_demo\windows\runner\flutter_window.cpp
[ +100 ms]   generated_plugin_registrant.cc
[ +448 ms]   flutter_window.cpp
[ +993 ms]   코드를 생성하고 있습니다...
[ +107 ms] ResourceCompile:
[   +1 ms]   모든 출력이 최신 상태입니다.
[  +28 ms] Link:
[        ]   C:\Program Files (x86)\Microsoft Visual
Studio\2019\Community\VC\Tools\MSVC\14.29.30133\bin\HostX64\x64\link.exe /ERRORREPORT:QUEUE
/OUT:"C:\Users\robin\Documents\workspace\twoeat_flutter_demo\build\windows\runner\Debug\twoeat_flutter_demo.ex
e" /INCREMENTAL /ILK:"twoeat_flutter_demo.dir\Debug\twoeat_flutter_demo.ilk" /NOLOGO
..\flutter\Debug\flutter_wrapper_app.lib
C:\Users\robin\Documents\workspace\twoeat_flutter_demo\windows\flutter\ephemeral\flutter_windows.dll.lib
kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib
advapi32.lib /MANIFEST /MANIFESTUAC:"level='asInvoker' uiAccess='false'" /manifest:embed
/manifestinput:C:\Users\robin\Documents\workspace\twoeat_flutter_demo\windows\runner\runner.exe.manifest
/DEBUG
/PDB:"C:/Users/robin/Documents/workspace/twoeat_flutter_demo/build/windows/runner/Debug/twoeat_flutter_demo.pd
b" /SUBSYSTEM:WINDOWS /TLBID:1 /DYNAMICBASE /NXCOMPAT
/IMPLIB:"C:/Users/robin/Documents/workspace/twoeat_flutter_demo/build/windows/runner/Debug/twoeat_flutter_demo
.lib" /MACHINE:X64  /machine:x64 twoeat_flutter_demo.dir\Debug\Runner.res
[   +1 ms]   twoeat_flutter_demo.dir\Debug\flutter_window.obj
[        ]   twoeat_flutter_demo.dir\Debug\main.obj
[        ]   twoeat_flutter_demo.dir\Debug\run_loop.obj
[        ]   twoeat_flutter_demo.dir\Debug\utils.obj
[        ]   twoeat_flutter_demo.dir\Debug\win32_window.obj
[        ]   twoeat_flutter_demo.dir\Debug\generated_plugin_registrant.obj
[ +277 ms]   twoeat_flutter_demo.vcxproj ->
C:\Users\robin\Documents\workspace\twoeat_flutter_demo\build\windows\runner\Debug\twoeat_flutter_demo.exe
[   +8 ms] FinalizeBuildStatus:
[        ]   "twoeat_flutter_demo.dir\Debug\twoeat_f.FA8FDEFE.tlog\unsuccessfulbuild" 파일을 삭제하고 있습니다
.
[        ]   "twoeat_flutter_demo.dir\Debug\twoeat_f.FA8FDEFE.tlog\twoeat_flutter_demo.lastbuildstate"에
연결(touching)하고 있습니다.
[   +1 ms]
"C:\Users\robin\Documents\workspace\twoeat_flutter_demo\build\windows\runner\twoeat_flutter_demo.vcxproj"
프로젝트를 빌드했습니다(기본 대상).
[   +8 ms] InitializeBuildStatus:
[   +1 ms]   "AlwaysCreate"이(가) 지정되었기 때문에 "x64\Debug\ALL_BUILD\ALL_BUILD.tlog\unsuccessfulbuild"을(
를) 만들고 있습니다.
[   +2 ms] CustomBuild:
[        ]   모든 출력이 최신 상태입니다.
[   +9 ms] FinalizeBuildStatus:
[        ]   "x64\Debug\ALL_BUILD\ALL_BUILD.tlog\unsuccessfulbuild" 파일을 삭제하고 있습니다.
[        ]   "x64\Debug\ALL_BUILD\ALL_BUILD.tlog\ALL_BUILD.lastbuildstate"에 연결(touching)하고 있습니다.
[        ] "C:\Users\robin\Documents\workspace\twoeat_flutter_demo\build\windows\ALL_BUILD.vcxproj" 프로젝트를
빌드했습니다(기본 대상).
[   +7 ms] InitializeBuildStatus:
[        ]   "AlwaysCreate"이(가) 지정되었기 때문에 "x64\Debug\INSTALL\INSTALL.tlog\unsuccessfulbuild"을(를)
만들고 있습니다.
[  +82 ms] PostBuildEvent:
[   +1 ms]   setlocal
[        ]   "C:\Program Files (x86)\Microsoft Visual
Studio\2019\Community\Common7\IDE\CommonExtensions\Microsoft\CMake\CMake\bin\cmake.exe" -DBUILD_TYPE=Debug -P
cmake_install.cmake
[        ]   if %errorlevel% neq 0 goto :cmEnd
[        ]   :cmEnd
[        ]   endlocal & call :cmErrorLevel %errorlevel% & goto :cmDone
[        ]   :cmErrorLevel
[        ]   exit /b %1
[        ]   :cmDone
[        ]   if %errorlevel% neq 0 goto :VCEnd
[        ]   :VCEnd
[  +56 ms]   -- Install configuration: "Debug"
[  +37 ms]   -- Up-to-date:
C:/Users/robin/Documents/workspace/twoeat_flutter_demo/build/windows/runner/Debug/data/icudtl.dat
[   +1 ms]   -- Up-to-date:
C:/Users/robin/Documents/workspace/twoeat_flutter_demo/build/windows/runner/Debug/flutter_windows.dll
[   +9 ms]   -- Installing:
C:/Users/robin/Documents/workspace/twoeat_flutter_demo/build/windows/runner/Debug/data/flutter_assets
[   +1 ms]   -- Installing:
C:/Users/robin/Documents/workspace/twoeat_flutter_demo/build/windows/runner/Debug/data/flutter_assets/AssetMan
ifest.json
[   +2 ms]   -- Installing:
C:/Users/robin/Documents/workspace/twoeat_flutter_demo/build/windows/runner/Debug/data/flutter_assets/FontMani
fest.json
[   +2 ms]   -- Installing:
C:/Users/robin/Documents/workspace/twoeat_flutter_demo/build/windows/runner/Debug/data/flutter_assets/fonts
[   +2 ms]   -- Installing:
C:/Users/robin/Documents/workspace/twoeat_flutter_demo/build/windows/runner/Debug/data/flutter_assets/fonts/Ma
terialIcons-Regular.otf
[   +7 ms]   -- Installing:
C:/Users/robin/Documents/workspace/twoeat_flutter_demo/build/windows/runner/Debug/data/flutter_assets/kernel_b
lob.bin
[ +208 ms]   -- Installing:
C:/Users/robin/Documents/workspace/twoeat_flutter_demo/build/windows/runner/Debug/data/flutter_assets/NOTICES
[   +6 ms]   -- Installing:
C:/Users/robin/Documents/workspace/twoeat_flutter_demo/build/windows/runner/Debug/data/flutter_assets/packages
[   +1 ms]   -- Installing:
C:/Users/robin/Documents/workspace/twoeat_flutter_demo/build/windows/runner/Debug/data/flutter_assets/packages
/cupertino_icons
[   +1 ms]   -- Installing:
C:/Users/robin/Documents/workspace/twoeat_flutter_demo/build/windows/runner/Debug/data/flutter_assets/packages
/cupertino_icons/assets
[   +1 ms]   -- Installing:
C:/Users/robin/Documents/workspace/twoeat_flutter_demo/build/windows/runner/Debug/data/flutter_assets/packages
/cupertino_icons/assets/CupertinoIcons.ttf
[  +15 ms] FinalizeBuildStatus:
[        ]   "x64\Debug\INSTALL\INSTALL.tlog\unsuccessfulbuild" 파일을 삭제하고 있습니다.
[   +1 ms]   "x64\Debug\INSTALL\INSTALL.tlog\INSTALL.lastbuildstate"에 연결(touching)하고 있습니다.
[        ] "C:\Users\robin\Documents\workspace\twoeat_flutter_demo\build\windows\INSTALL.vcxproj" 프로젝트를
빌드했습니다(기본 대상).
[  +18 ms] 빌드했습니다.
[   +1 ms]     경고 0개
[        ]     오류 0개
[        ] 경과 시간: 00:00:10.72
[  +34 ms] Building Windows application... (completed in 11.7s)
[ +247 ms] Observatory URL on device: http://127.0.0.1:13264/HkOcQA5wzKY=/
[   +7 ms] Caching compiled dill
[  +40 ms] Connecting to service protocol: http://127.0.0.1:13264/HkOcQA5wzKY=/
[ +179 ms] Launching a Dart Developer Service (DDS) instance at http://127.0.0.1:0, connecting to VM service
at http://127.0.0.1:13264/HkOcQA5wzKY=/.
[ +101 ms] DDS is listening at http://127.0.0.1:13268/joKeVxVyuT8=/.
[  +88 ms] Successfully connected to service protocol: http://127.0.0.1:13264/HkOcQA5wzKY=/
[  +20 ms] DevFS: Creating new filesystem on the device (null)
[  +24 ms] DevFS: Created new filesystem on the device
(file:///C:/Users/robin/AppData/Local/Temp/twoeat_flutter_demo5be71424/twoeat_flutter_demo/)
[   +3 ms] Updating assets
[  +92 ms] Manifest contained wildcard assets. Inserting missing file into build graph to force rerun. for
more information see #56466.
[   +6 ms] Syncing files to device Windows...
[   +2 ms] <- reset
[   +1 ms] Compiling dart to kernel with 0 updated files
[   +6 ms] <- recompile package:twoeat_flutter_demo/main.dart fa69b64c-a8d3-48cd-bf05-00d4a4be28db
[   +3 ms] <- fa69b64c-a8d3-48cd-bf05-00d4a4be28db
[ +114 ms] Updating files.
[   +1 ms] DevFS: Sync finished
[   +2 ms] Syncing files to device Windows... (completed in 131ms)
[   +1 ms] Synced 0.0MB.
[   +1 ms] <- accept
[   +7 ms] Connected to _flutterView/0x24ed74b51e0.
[   +3 ms] Flutter run key commands.
[   +3 ms] r Hot reload. 
[   +2 ms] R Hot restart.
[   +1 ms] h Repeat this help message.
[   +1 ms] d Detach (terminate "flutter run" but leave application running).
[   +1 ms] c Clear the screen
[   +1 ms] q Quit (terminate the application on the device).
[   +1 ms]  Running with sound null safety 
[   +1 ms] An Observatory debugger and profiler on Windows is available at:
http://127.0.0.1:13268/joKeVxVyuT8=/
[+1005 ms] The Flutter DevTools debugger and profiler on Windows is available at:
                    http://127.0.0.1:9103?uri=http%3A%2F%2F127.0.0.1%3A13268%2FjoKeVxVyuT8%3D%2F


```

```
Analyzing twoeat_flutter_demo...
No issues found! (ran in 2.1s)

```

```
[✓] Flutter (Channel stable, 2.2.3, on Microsoft Windows [Version 10.0.19043.1165], locale ko-KR)
    • Flutter version 2.2.3 at C:\Users\robin\Documents\flutter
    • Framework revision f4abaa0735 (7 weeks ago), 2021-07-01 12:46:11 -0700
    • Engine revision 241c87ad80
    • Dart version 2.13.4

[✗] Android toolchain - develop for Android devices
    ✗ Unable to locate Android SDK.
      Install Android Studio from: https://developer.android.com/studio/index.html
      On first launch it will assist you in installing the Android SDK components.
      (or visit https://flutter.dev/docs/get-started/install/windows#android-setup for detailed instructions).
      If the Android SDK has been installed to a custom location, please use
      `flutter config --android-sdk` to update to that location.


[✓] Chrome - develop for the web
    • Chrome at C:\Program Files\Google\Chrome\Application\chrome.exe

[✓] Visual Studio - develop for Windows (Visual Studio Community 2019 16.11.1)
    • Visual Studio at C:\Program Files (x86)\Microsoft Visual Studio\2019\Community
    • Visual Studio Community 2019 version 16.11.31613.86
    • Windows 10 SDK version 10.0.19041.0

[!] Android Studio (not installed)
    • Android Studio not found; download from https://developer.android.com/studio/index.html
      (or visit https://flutter.dev/docs/get-started/install/windows#android-setup for detailed instructions).

[✓] IntelliJ IDEA Ultimate Edition (version 2021.2)
    • IntelliJ at C:\Users\robin\AppData\Local\JetBrains\Toolbox\apps\IDEA-U\ch-0\212.4746.92
    • Flutter plugin version 59.0.4
    • Dart plugin version 212.4746.57

[✓] VS Code (version 1.59.0)
    • VS Code at C:\Users\robin\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension can be installed from:
      🔨 https://marketplace.visualstudio.com/items?itemName=Dart-Code.flutter

[✓] Connected device (3 available)
    • Windows (desktop) • windows • windows-x64    • Microsoft Windows [Version 10.0.19043.1165]
    • Chrome (web)      • chrome  • web-javascript • Google Chrome 92.0.4515.159
    • Edge (web)        • edge    • web-javascript • Microsoft Edge 92.0.902.78

! Doctor found issues in 2 categories.
```

</details>

Edit by @cbracken 2022-01-11: It's been pointed out that this has become a sort of general catch-all for Win32 Korean input issues. This issue and issues such as https://github.com/flutter/flutter/issues/96209 will almost certainly have the same fix, and were de-duplicated to keep the tracking on the work in one place.

**Comment by TahaTesser** [comment]
Hi @yunu7067 
Thanks for filing the issue

> If I press space only once, a key such as Ctrl + A or Backspace will not work. 

I can reproduce this on `stable, 2.2.3`, however, it doesn't seem on the latest `beta, 2.5.0-5.2.pre`


Can you please upgrade to the latest  `beta` and try again
Execute the following commands to upgrade
`flutter channel beta`
`flutter upgrade --force`
`flutter doctor -v'

Thank you

**Comment by yunu7067** [comment]
(I apologize in advance for the lack of meaning, because I rely on a translator.)

Thanks for the quick reply.  I upgraded flutter to beta channel and checked it again.
The problem was solved, but it was modified differently than I expected.


![test2](https://user-images.githubusercontent.com/57068092/130794171-abd4329f-256b-4fd5-b78a-9d0c6cd18404.gif)

In the picture above, You can see that the input behavior of Windows Notepad and Flutter is different.
I can't enter any other characters, except for the Hangul alphabet, until the character is complete. For example, if I want to insert one space in the middle of typing in flutter, I have to press the space twice. Normal Korean input in Windows, just press it once.

**Comment by gspencergoog** [comment]
cc @justinmc @cbracken

**Comment by justinmc** [comment]
I don't know much about Korean so it's hard for me to tell for sure, but I tried this out on my Mac and it seemed to work correctly.  I'll plan to try this out on my Windows machine next time I've got that up and running.

**Comment by donaldu12** [comment]
I'm having the same problem. 
When typing with hardware keyboard in Korean on a text field, have to press the space bar twice to insert space between words. 
I only tested on Flutter desktop(windows) with hardware keyboard, but I haven't tested on Linux and MacOS yet

**Comment by sclee15** [comment]
The situation in Linux is much worse. It cannot type properly with iBus (ubuntu)

**Comment by smok95** [comment]
I'm having the same problem.

**Comment by cbracken** [comment]
I can reproduce this locally. This appears to be a regression that must have landed sometime during the text input refactoring that landed over the summer. I'm bumping priority since Korean text input is seriously impacted by this.

**Comment by cbracken** [comment]
> The situation in Linux is much worse. It cannot type properly with iBus (ubuntu)

@sclee15 please file a separate issue for the behaviour of Korean input on Linux; the fixes will be platform-specific so we'll want to track them in separate issues. Thanks for reporting.

**Comment by cbracken** [comment]
Some notes to myself or anyone else looking into this. I've instrumented the Win32 event loop to output the composition and result strings.

When I type "banana" (바나나), I get the following Win32 events (one for each keypress) 
Keypress: ㅂ (q)
[+13314 ms] Composition string: ㅂ
Flutter textfield: ㅂ (composing range: 0,1; selection: 0,0) **correct**

Keypress: ㅏ (k)
[+1275 ms] Composition string: 바
Flutter textfield: 바 (composing range: 0,1; selection: 0,0) **correct**

Keypress: ㄴ (s)
[+3983 ms] Composition string: 반
Flutter textfield: 반 (composing range: 0,1; selection: 0,0) **correct**

Keypress: ㅏ (k)
[+1368 ms] Result string: 바
[   +1 ms] Composition string: 나
Flutter textfield: 바나 (composing range: 1,2; selection: 1,1) **correct*

Keypress: ㄴ (s)
[+8717 ms] Composition string: 난
Flutter textfield: 난난바 (composing range: 0,0; selection: 0,0) **incorrect**

Keypress: ㅏ (k) 
[+2546 ms] Result string: 나
[   +1 ms] Composition string: 나
Flutter textfield: 나나나난난바 (composing range: 0,0; selection: 1,1) **incorrect**

Keypress: (space)
[+32096 ms] Result string:
Flutter textfield: 나나나난난나 (composing range: 0,0, selection: 2,2) **incorrect**

**Comment by cbracken** [comment]
It looks like we're getting a bad composing range from the framework at some point. Here's the above sequence with a bit more instrumentation.

+4954 ms] ===== Composition string: 'ㅂ' cursor_pos: 0
[   +2 ms] TextInputModel::AddText(u16string)
[   +1 ms] Current text: '' composing: 0,0 selection: 0,0
[   +1 ms] Deleting selection
[   +1 ms] Updated text: '' composing: 0,0 selection: 0,0
[   +1 ms] Deleting composing range.
[   +1 ms] Updated text: '' composing: 0,1 selection: 0,0
[   +1 ms] Inserting to composing range at pos: 0
[   +1 ms] New text: 'ㅂ' composing: 0,1 selection: 1,1
[+1448 ms] ===== Composition string: '바' cursor_pos: 0
[   +1 ms] TextInputModel::AddText(u16string)
[   +1 ms] Current text: 'ㅂ' composing: 0,1 selection: 0,0
[   +1 ms] Deleting selection
[   +1 ms] Updated text: 'ㅂ' composing: 0,1 selection: 0,0
[   +1 ms] Deleting composing range.
[   +1 ms] Updated text: '' composing: 0,1 selection: 0,0
[   +1 ms] Inserting to composing range at pos: 0
[   +1 ms] New text: '바' composing: 0,1 selection: 1,1
[ +957 ms] ===== Composition string: '반' cursor_pos: 0
[   +1 ms] TextInputModel::AddText(u16string)
[   +1 ms] Current text: '바' composing: 0,1 selection: 0,0
[   +1 ms] Deleting selection
[   +1 ms] Updated text: '바' composing: 0,1 selection: 0,0
[   +1 ms] Deleting composing range.
[   +1 ms] Updated text: '' composing: 0,1 selection: 0,0
[   +1 ms] Inserting to composing range at pos: 0
[   +1 ms] New text: '반' composing: 0,1 selection: 1,1
[ +789 ms] ===== Result string: '바' cursor_pos: 0
[   +1 ms] TextInputModel::AddText(u16string)
[   +1 ms] Current text: '반' composing: 0,1 selection: 0,0
[   +1 ms] Deleting selection
[   +1 ms] Updated text: '반' composing: 0,1 selection: 0,0
[   +1 ms] Deleting composing range.
[   +1 ms] Updated text: '' composing: 0,1 selection: 0,0
[   +1 ms] Inserting to composing range at pos: 0
[   +1 ms] New text: '바' composing: 0,1 selection: 1,1
[   +1 ms] COMMIT
[   +1 ms] Updating selection
[   +1 ms] Current text: '바' composing: 1,1 selection: 1,1
[   +1 ms] ===== Composition string: '나' cursor_pos: 0
[   +1 ms] TextInputModel::AddText(u16string)
[   +1 ms] Current text: '바' composing: 1,1 selection: 1,1
[   +1 ms] Deleting selection
[   +1 ms] Updated text: '바' composing: 1,1 selection: 1,1
[   +1 ms] Deleting composing range.
[   +1 ms] Updated text: '바' composing: 1,2 selection: 1,1
[   +1 ms] Inserting to composing range at pos: 1
[   +1 ms] New text: '바나' composing: 1,2 selection: 2,2

Everything is looking fine up to here. But suddenly we get an update from the framework that resets the composing range.

[   +1 ms] ## Got composing -1,-1 from FW
[   +1 ms] END COMPOSING. composing: 0,0

I *suspect* this might be somehow related to the fact that we perform a commit above since the IME commits the first character and starts composing the second. If so, then perhaps the framework assumes that there's no longer a composing range and sends a state update back (asynchronously) with a -1,-1 composing range. After this, things are in bad shape:

[+20949 ms] ===== Composition string: '난' cursor_pos: 0
[   +2 ms] TextInputModel::AddText(u16string)
[   +1 ms] Current text: '바' composing: 0,0 selection: 0,0
[   +1 ms] Deleting selection
[   +1 ms] Updated text: '바' composing: 0,0 selection: 0,0
[   +1 ms] Inserting to composing range at pos: 0
[   +1 ms] Updated text: '난바' composing: 0,0 selection: 1,1
[+5300 ms] ## ComposeEndHook
[   +1 ms] COMMIT
[   +1 ms] Updating selection
[   +1 ms] Current text: '난난바d�' composing: 1,1 selection: 1,1
[   +1 ms] END COMPOSING. composing: 0,0

**Comment by cbracken** [comment]
Also /cc @dkwingsmt in case he's got thoughts on this.

**Comment by HansMuller** [comment]
May be related: https://github.com/flutter/flutter/issues/96028

**Comment by justinmc** [comment]
@cbracken Sorry for the late response here.  Feel free to ping me if you're still digging into the framework part of this.

Some platforms send "ACK" messages between the framework and engine immediately after receiving an update to the text editing state.  I can't remember if Windows does this or not, but it might be something to watch out for in the case of the IME sending immediate back to back updates like you mention above.  If the first update causes an ACK to come back asynchronously later, it could arrive after the second IME update and overwrite it.

**Comment by cbracken** [comment]
A few updates.

Bisection reveals that this regression was introduced in https://github.com/flutter/flutter/pull/90211 (which is a re-land of https://github.com/flutter/flutter/pull/89327). It appears to be the result of the fact that single-line text fields include a `TextInputFormatter` (to prevent newlines) which fires updates over the text input channel back to the engine where previously we didn't send a message back immediately after the commit of a character.  Multi-line text fields do not exhibit this bug.

Here's what seems to be happening:
1. User presses ㅂ (q).
   Engine->Framework: ㅂ (composing range: 0,1; selection: 0,0)
2. User presses ㅏ (k).
   Engine->Framework: 바 (composing range: 0,1; selection: 0,0)
3. User presses ㄴ (s).
   Engine->Framework: 반 (composing range: 0,1; selection: 0,0)

So far everything is fine.

4. User presses ㅏ (k).

This triggers the completion of the first character and the addition of a second one, causing three engine->framework updates, one after the other, synchronously.

4.a GCS_RESULTSTR event: Engine->Framework: 바 (composing: 0,1 selection: 0,0)
4.b. Commit: Engine->Framework: 바 (composing: 1,1 selection: 1,1)
4.c. GCS_COMPSTR event: Engine->Framework: 바나 (composing: 1,2 selection: 1,1)

In response to the message sent to the framework in 4.b, the TextField's FilteringTextInputFormatter triggers an updated TextEditingValue that gets sent back to the engine. However, this message is already out of date, since message 4.c has already been pushed from the engine onto the text input channel in the meantime.

5. Framework->Engine: 바 (composing: -1,-1 selection: 1,1)

This causes the composing range in the engine state to get reset back to an invalid range, so the next keypress results in erroneous behaviour.

**Comment by cbracken** [comment]
One more update. This is fixed by flutter/engine#30805.

There is however, a bug related to enter key behaviour, though it's general and not specific to Korean input. It can be found here:
https://github.com/flutter/flutter/issues/93373
